### PR TITLE
Generate Span object from SpanOperation

### PR DESCRIPTION
**Pull request by delner**
Expanding upon the implementation of `SpanOperation`, this pull request seeks to change its behavior from wrapping and mutating a `Span`, to building a copy of a `Span` only on `SpanOperation#finish`. In effect, the `SpanOperation` acts as a builder or factory pattern.

The main motivation to do this is to:

1. Simplify `Span` objects into basic structs suitable for wire serialization
2. Prevent their mutation after completion so that instrumentation does not produce race conditions or other unexpected behavior by modifying a completed `Span` by reference. (Eventually, we'd like to make `Span` outright immutable, but for now this is an intermediate step.)
3. Reduce the references held on the `Tracer` side of instrumentation, to reduce the likelihood of memory leaks.

In doing so, much of the timing behavior (`start`, `stop`) and `Span` state have migrated to `SpanOperation`. While refining `SpanOperation`, I also decided to privatize a few functions (`context=`, `parent=`, `events`, `span`) to dissuade public use, as there are plans to refactor or eliminate them entirely in the future. Each of these should have sufficient alternatives through other APIs we prefer users to use.

### CODE REVIEW

File: lib/ddtrace/tracer.rb [hunk 1]
```diff
@@ -383,9 +402,7 @@ def subscribe_on_error(span_op, error_handler)
     # Records the span (& its context)
     def record_span(span_op)
       begin
-        span_op.service ||= default_service
```
Comment (marcotc) [inline lib/ddtrace/tracer.rb:386]: Is setting the `default_service` as a fallback after when the span is ready to be flushed done elsewhere now?

Comment (delner) [reply lib/ddtrace/tracer.rb:386]: Hmmmm....

So the answer is mostly yes, which is why I removed this. See `Tracer#build_span_options`: it decides and specifies the `service` name before it builds the `SpanOperation`. A `Tracer` spec that tests for `default_service` confirms this behavior is working correctly.

The "maybe no" part of this is that its possible for a `SpanOperation` to be manually created, not have a `service` specified, then flushed through a `Tracer#record`. In reality, this is never done in practice, or recommended. The `Tracer` only knows how to flush `Context` objects anyways, not individual `Span`s or `SpanOperation`s, so it's basically impossible.

All that said, I think this should be a safe change.

**Review by marcotc — APPROVED** [review]

File: lib/ddtrace/span.rb [hunk 2]
```diff
@@ -21,6 +14,8 @@ module Datadog
   #
   # rubocop:disable Metrics/ClassLength
   class Span
```
Comment (ivoanjo) [inline lib/ddtrace/span.rb:16]: Hmm. I wonder if for compatibility, we should rename `SpanOperation` => `Span` and `Span` => `CompletedSpan` or something like that?

It looks to me like, after all these changes, `Span` is essentially unrecognizable, and `SpanOperation` was what we always called span, so perhaps recovering that name would be helpful for others as well?

File: lib/ddtrace/span.rb [hunk 3]
```diff
@@ -34,224 +29,135 @@ class Span
...
     #
     # * +service+: the service name for this span
     # * +resource+: the resource this span refers, or +name+ if it's missing.
     #     +nil+ can be used as a placeholder, when the resource value is not yet known at +#initialize+ time.
-    # * +span_type+: the type of the span (such as +http+, +db+ and so on)
+    # * +type+: the type of the span (such as +http+, +db+ and so on)
     # * +parent_id+: the identifier of the parent span
     # * +trace_id+: the identifier of the root span for this trace
+    # TODO: Remove span_type
     def initialize(
```
Comment (ivoanjo) [inline lib/ddtrace/span.rb:66]: It doesn't seem that we're very far from being able to call `.freeze` in the constructor. Would that be something we eventually want to get to?

File: docs/GettingStarted.md [hunk 4]
```diff
@@ -327,9 +327,8 @@ To trace these operations, you can trace code asynchronously by calling `Datadog
 ```ruby
 # Some instrumentation framework calls this after an event finishes...
 def db_query(start, finish, query)
-  span = Datadog.tracer.trace('database.query')
+  span = Datadog.tracer.trace('database.query', start_time: start)
   span.resource = query
```
Comment (ivoanjo) [inline docs/GettingStarted.md:331]: Minor: Could we also pass in the `resource` when starting the span?

File: lib/ddtrace/tagging/analytics.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+          Datadog::Analytics.set_sample_rate(self, value)
+        when Ext::Analytics::TAG_SAMPLE_RATE
+          Datadog::Analytics.set_sample_rate(self, value)
+        else
+          super if defined?(super)
+        end
+      end
+    end
+  end
+end
```
Comment (ivoanjo) [inline lib/ddtrace/tagging/analytics.rb:22]: Are `Datadog::Tagging::Analytics` and `Datadog::Tagging::ManualTracing` used independently of `Metadata` anywhere?

If not, I would suggest actually merging them back into `Metadata`: With the current arrangement, when someone that wants to understand and follow the current behavior of `Span#set_tag` they need to find 3 different methods across 3 different files (none of which is actually `span.rb`), which is really hard to follow.

File: lib/ddtrace/pipeline.rb [hunk 6]
```diff
@@ -30,6 +30,8 @@ def self.processors=(value)
 
     def self.apply_processors!(trace)
       result = @processors.inject(trace) do |current_trace, processor|
+        next nil if current_trace.nil?
```
Comment (ivoanjo) [inline lib/ddtrace/pipeline.rb:33]: Minor: `next unless current_trace`? :)

**Review by ivoanjo — COMMENTED** [review]
I realize I'm already a bit late, but left a few notes that seemed relevant :)
